src/haversine.jl

```diff
@@ -61,4 +62,5 @@ end
 
 spherical_angle(x, y) = SphericalAngle()(x, y)
 
-result_type(::Union{Haversine, SphericalAngle}, ::Type, ::Type) = Float64
+result_type(d::Haversine{T1}, ::Type{T2}, ::Type{T3}) where {T1<:Number,T2<:Number,T3<:Number} = promote_type(T1, T2, T3)
```

Comment:
I'm not sure this is correct. If all types are `<:Integer`, then the result is still some float because you're computing `[a]sin[d]`.

Reply:
True! Make it `float(...)`?

Honestly wasn't aware of this `return_type` when I made the changes. Could we just use `Core.Compiler.return_type(d, map(typeof, (x, y)))` or is that a bad idea?

Reply:
I don't think there is anything wrong with the default `Float64` because that's what you will get from the `sin` computations anyway. If at some point somebody finds that `Float64` is too narrow for certain input types, then we can still debug that case.

Reply:
> I don't think there is anything wrong with the default Float64 because that's what you will get from the sin computations anyway.

Maybe I misunderstand you but that's not the true, i.e. `sin(::Float32)::Float32` is true. This is exactly why I made this PR:) Even the current implementation is strange since by default `Haversine` uses `Float32` for the radius and so `(::Haversine)(x::Vector{Float32}, y::Vector{Float32})` returns `Float32` _but_ if you use it with `pairwise` it returns `eltype(...) = Float64` because of the `return_type` above.

Reply:
Honestly, since the computations of the distances are seemingly "simple", you might be able to get away with using `Core.Compiler.return_type` to infer `result_type` for all distances instead of the hardcoding `result_type`.

Reply:
> so `(::Haversine)(x::Vector{Float32}, y::Vector{Float32})` returns `Float32`

True, I wasn't aware of that. Then yes, hardcoding `Float64` is not the right thing to do. Then I guess your `float` idea is the way to go. IIRC, then usage of `Core.Compile.return_type` is absolutely not recommended.

Reply:
> Then I guess your float idea is the way to go.

Cool, I'll do that then :+1: 

> IIRC, then usage of Core.Compile.return_type is absolutely not recommended.

See I've heard the same, but IIRC the reason is that it might silently return `Any` but was thinking that since the functions used here are so "simple" it shouldn't be much of a problem. And in the cases where it is, then maybe `Any` is the best we can do?

But confirming/denying that hypothesis would require a bit more and so I don't think it's suitable for this PR anyways:)

Reply:
We still need the `float` of the promote_type, don't we?

Reply:
You could check with integer arguments.

Reply:
Oh good catch; forgot about that! Doneso.